src/vm/Components/Assembler/Assembler.java

```diff
@@ -0,0 +1,208 @@
+    private Map<String, Integer> symbolTable;
+
+    public Assembler() {
+        addressCounter = 0;
+        lineCounter = 1;
+        instructionsMap = new HashMap<>();
+        symbolTable = new HashMap<>();
+        setInstructionData();
+        assemble();
+    }
+
+    public void assemble() {
+        try {
+            firstPass();
+        } catch (FileNotFoundException e) {
+            e.printStackTrace();
+            System.out.println("Erro no processo do primeiro passo do montador!");
+        } catch (AssemblerException e) {
+            return;
+        } finally {
+            System.out.println(symbolTable);
+        }
+
+        try {
+            secondPass();
+        } catch (FileNotFoundException e) {
+            e.printStackTrace();
+            System.out.println("Erro no processo do segundo passo do montador!");
+        }
+    }
+
+    private void firstPass() throws FileNotFoundException, AssemblerException {
+        Reader reader = new Reader("MASMAPRG.ASM");
+        String line = reader.readLine();
+
+        // retira comentarios
+        line = line.split(";")[0];
+
+        // endereco sera zero enquanto desconhecido
+        int address = 0;
+        while (line != null) {
+
+            if (line.length() > 80) {
+                throw new AssemblerException("Assembler Exception: Invalid input at line " + lineCounter + ". Lines should have a maximum length of 80 characters.");
+            }
+
+            String [] params = line.split(" ");
+            if (params.length > 4) {
+                throw new AssemblerException("Assembler Exception: Syntax error at line " + lineCounter + ". Too many parameters.");
+            }
+
+            for (String param : params) {
+
+                param = param.trim();
+
+                // caso seja constante
+                if (param.replace("@", "").matches("^\\d+$")) {
+                    // testa se pode ser representado com 16 bits
+                    if (Long.valueOf(param) > 65536) {
+                        throw new AssemblerException("Assembler Exception: Value (" + param + ") out of bounds at line " + lineCounter + ". Constant is larger than 16 bits.");
+                    }
+                // caso nao seja constante, testa se o simbolo comeca com uma letra
+                } else if (!param.matches("[a-zA-Z_][a-zA-Z0-9_]*")) {
+                    throw new AssemblerException("Assembler Exception: Invalid argument (" + param + ") at line " + lineCounter + ". Symbols should start with a letter.");
+                }
+
+                // caso o parametro seja uma instrucao
+                if (instructionsMap.containsKey(param)) {
+                    addressCounter += instructionsMap.get(param).getInstructionSize();
+                // caso o parametro seja um simbolo
+                } else if (!instructionsMap.containsKey(param) && param.matches("[a-zA-Z_][a-zA-Z0-9_]*")) {
+
+                    if (param.length() > 8) {
```

Comment:
mesma de cima

Reply:
resolved in commit https://github.com/guivahl/programacao-de-sistemas/commit/7fb99ebb5fb76f7c06d91c084deb1fb094fa2210